### ISSUE TITLE
Users of PHP Imagick need to recompile when upgrading ImageMagick.

### ISSUE THREAD

**Issue by Danack** [Homebrew/homebrew-php#3919]
*Note - this is separate from https://github.com/Homebrew/homebrew-php/pull/3913 which would also be required to use a version of Imagick that compiles against ImageMagick 7*

Hello there,

I'm the maintainer of PHP Imagick.

I thought I'd drop a note to clarify an issue Homebrew users seem to be having with PHP Imagick, to avoid traffic elsewhere*

Imagick is a very thin wrapper around the ImageMagick library. When Imagick is compiled it looks at the ImageMagick files it is being compiled against, and uses values and definitions of structs to compile Imagick according to that specific version of ImageMagick that is on the users screen.

This means that Imagick is only likely to work against the exact version it was compiled against, and definitely won't work when it is compiled against ImageMagick 6 but ImageMagick 7 is installed on the users system.

What seems to be happening is that users are:

1) Installing ImageMagick 6
2) Installing PHP Imagick
3) Upgrading ImageMagick to 7
4) Seeing that PHP Imagick is no longer working.

For example:

https://github.com/Homebrew/homebrew-php/issues/3916
> After update imagick to imagemagick-7.0.4-4
> Library not loaded: /usr/local/opt/imagemagick/lib/libMagickWand-6.Q16.3.dylib

This user has definitely upgraded to 7, but apparently Imagick is still looking for the version 6 library.

https://github.com/Homebrew/homebrew-php/issues/3912
> Library not loaded: /usr/local/opt/imagemagick/lib/libMagickWand-6.Q16.3.dylib

I can't tell if this person has upgraded to IM7, but Imagick is complaining about not being able to find the version 6 library.

https://github.com/ImageMagick/ImageMagick/issues/353
This user, who certainly makes me feel warm and fuzzy, and not all like deleting the whole repository and walking away, is lost when it comes to what is happening with regards to making PHP Imagick work with ImageMagick 7.

tl:dr - it's apparently not obvious for users that upgrading ImageMagick to 7 requires doing something to Imagick. I'm not sure what capabilities HomeBrew has to enforce downstream projects get recompiled when upstream projects change, but even doing something simple like displaying a message at the end of the ImageMagick compilation process saying "NOW REINSTALL IMAGICK, IF YOU USE IT", might be of some help.

cheers
Dan
Ack

**Comment by hellopablo** [comment]
Hey @Danack - I just tried doing as you suggested and still having issues. I have IM7 installed, but reinstalling `php56-imagick` still seems to want to load the IM6 libraries. Am I missing something obvious? I even tried completely removing IM and imagick, then `brew install php56-imagick` and letting brew handle the IM dependancy, but nada.

```
➜  php -v
PHP Warning:  PHP Startup: Unable to load dynamic library '/usr/local/opt/php56-imagick/imagick.so' - dlopen(/usr/local/opt/php56-imagick/imagick.so, 9): Library not loaded: /usr/local/opt/imagemagick/lib/libMagickWand-6.Q16.3.dylib
  Referenced from: /usr/local/opt/php56-imagick/imagick.so
  Reason: image not found in Unknown on line 0

Warning: PHP Startup: Unable to load dynamic library '/usr/local/opt/php56-imagick/imagick.so' - dlopen(/usr/local/opt/php56-imagick/imagick.so, 9): Library not loaded: /usr/local/opt/imagemagick/lib/libMagickWand-6.Q16.3.dylib
  Referenced from: /usr/local/opt/php56-imagick/imagick.so
  Reason: image not found in Unknown on line 0
PHP 5.6.29 (cli) (built: Jan 17 2017 17:43:44) 
Copyright (c) 1997-2016 The PHP Group
Zend Engine v2.6.0, Copyright (c) 1998-2016 Zend Technologies
```

As expected, looking in `/usr/local/opt/imagemagick/lib` shows the IM7 libraries.

**Comment by Danack** [comment]
@hellopablo This is a separate additional thing from #3913 which would be required to use a version of Imagick that compiles against ImageMagick 7

**Comment by PopkornKevin** [comment]
Same issue as @hellopablo, fresh setup install on a new mac.

**Comment by hellopablo** [comment]
Thanks @Danack - For what it's worth, here's the output of the rebuild from source. Similar errors to @PopkornKevin 

```
brew reinstall php56-imagick --build-from-source
==> Reinstalling homebrew/php/php56-imagick 
==> Downloading https://pecl.php.net/get/imagick-3.3.0.tgz
######################################################################## 100.0%
==> /usr/local/opt/php56/bin/phpize
==> ./configure --prefix=/usr/local/Cellar/php56-imagick/3.3.0_6 --with-php-config=/usr/local/opt/php56/bin/php-config --with-imagick=/usr/local/opt/imagemagick
==> make
Last 15 lines from /Users/pablo/Library/Logs/Homebrew/php56-imagick/03.make:
/private/tmp/php56-imagick-20170118-58974-1ud81qs/imagick-3.3.0/imagick_helpers.c:1140:55: error: use of undeclared identifier 'Rec601LumaColorspace'
        IMAGICK_REGISTER_CONST_LONG("COLORSPACE_REC601LUMA", Rec601LumaColorspace);
                                                             ^
/private/tmp/php56-imagick-20170118-58974-1ud81qs/imagick-3.3.0/imagick_helpers.c:1141:55: error: use of undeclared identifier 'Rec709LumaColorspace'
        IMAGICK_REGISTER_CONST_LONG("COLORSPACE_REC709LUMA", Rec709LumaColorspace);
                                                             ^
/private/tmp/php56-imagick-20170118-58974-1ud81qs/imagick-3.3.0/imagick_helpers.c:1165:61: error: use of undeclared identifier 'ConstantVirtualPixelMethod'
        IMAGICK_REGISTER_CONST_LONG("VIRTUALPIXELMETHOD_CONSTANT", ConstantVirtualPixelMethod);  /* deprecated */
                                                                   ^
/private/tmp/php56-imagick-20170118-58974-1ud81qs/imagick-3.3.0/imagick_helpers.c:1266:53: error: use of undeclared identifier 'BicubicInterpolatePixel'
        IMAGICK_REGISTER_CONST_LONG("INTERPOLATE_BICUBIC", BicubicInterpolatePixel);
                                                           ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make: *** [imagick_helpers.lo] Error 1

READ THIS: http://docs.brew.sh/Troubleshooting.html

These open issues may also help:
PHP56-imagick appears to be broken https://github.com/Homebrew/homebrew-php/issues/3916
```

**Comment by fdelaiglesia-mobaloo** [comment]
Same problem here as @hellopablo  and @PopkornKevin with same output.

After googling found out something similar happening with ruby solved by manually linking imagick@6 again to /usr/local/opt

I used:
ln /usr/local/Cellar/imagemagick/6.9.7-0/lib/* /usr/local/opt/imagemagick/lib/

restart apache

Fixed!!

**Comment by fdelaiglesia-mobaloo** [comment]
The fix using brew seems to be:
brew unlink imagemagick
brew install imagmagick@6 (keg only)
brew link imagemagick@6 —force

then restart php and apache

Thanks to @meljturner
https://github.com/Homebrew/homebrew-php/issues/3912

**Comment by captain-geek** [comment]
My personal fix for this issue was to manually install imagick-3.4.x manually using PECL instead of using the homebrew package ... I'm confused as to why the fix is not upgrading the imagick package referenced in the homebrew formula ... 3.3.0 -> 3.4.1

Version 3.4.1 is considered stable and compiles against both ImageMagick 6 & 7

**Comment by javian** [comment]
These issues should now have been corrected thanks to @alcaeus and thank you for your feedback. Sorry that it took a bit of time to get this merged but I've been busy with some other things in life.

**Comment by SMillerDev** [comment]
@cAstraea It still exists, check if you can find php56 itself because you might not have homebrew-php tapped.

**Comment by hardikim** [comment]
Fatal error: ImagickDemo\Config::__construct(): Failed opening required '/var/www/html/Imagick-demos-master/src/ImagickDemo/../../../clavis.php' (include_path='.:/usr/share/php:/usr/share/pear') in /var/www/html/Imagick-demos-master/src/ImagickDemo/Config.php on line 43 how this solved all composer and all i install but not working and this error accur

**Comment by philip** [comment]
Bug https://github.com/Homebrew/homebrew-php/issues/3842 also refers to this with the following solution (in my case, for PHP71): `brew reinstall -s php71-imagick`

**Comment by adrianbj** [comment]
Thanks for this @philip - I used that approach with php72 and it works great. OT, but I have also had to do this with opcache for the last few versions of php as well - a bit annoying, but at least it works.
